Sampling docstrings

michaelosthege opened the pull request:
Editing docstrings and type hints on internal methods to make the code easier to understand.

Based on this call tree, there are manly 3 routes through the sampling process:
![sampling](https://user-images.githubusercontent.com/5894642/72689946-6bf92c80-3b17-11ea-8b2a-c600c7ad3c1e.png)

Some arguments such as `trace` are passed down the call tree quite a lot before they are actually used. Therefore the `Parameters` sections are often very similar..

Given the Python 2 EOL we could put `, *,` in many places.

There's a TODO comment in at least one place where I think we're wasting CPU time.

michaelosthege committed:
add docstring to _choose_backend

michaelosthege committed:
docstring in singleprocess sampling call tree

michaelosthege committed:
docstrings in population sampling

michaelosthege committed:
more accurate docstring/comment on population sampling

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -588,6 +610,35 @@ def _sample_population(
+    chains : int
+        The total number of chains in the population
+    step : function
+        Step function (should be or contain a population step method)
```
docstring out of order! the arguments go in a slightly different way

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -635,6 +719,7 @@ def _sample(
         strace = None
         for it, (strace, diverging) in enumerate(sampling):
             if it >= skip_first:
+                # TODO: unused variable - unnecessary creation of MultiTrace
```
i agree this could probably go. only thing i can think of is if `MultiTrace.__init__` does something disgusting that changes `strace` in place, but that would also be terrible.

ColCarroll replied:
(and, to be clear, I don't think it does that, but I also did not check that it does not)

michaelosthege replied:
I just removed the line already. I checked and the `MultiTrace.__init__` did not have a relevance here.

ColCarroll commented on pymc3/sampling.py:
```diff
@@ -1044,6 +1193,27 @@ def _iter_population(draws, tune, popstep, steppers, traces, points):
 
 
 def _choose_backend(trace, chain, shortcuts=None, **kwds):
+    """Selects or creates a trace for a particular chain.
```
```suggestion
    """Selects or creates a trace backend (NDArray, Text, etc) for a particular chain.
```

ColCarroll reviewed (approved):
this looks great. I think we should merge when you're happy, then delete that one line as a separate PR.

michaelosthege committed:
incorporate review comment

Co-Authored-By: Colin <ColCarroll@users.noreply.github.com>

michaelosthege committed:
correct order of arguments

michaelosthege committed:
take out creation of unused MultiTrace

michaelosthege committed:
type hints and docstring for _mp_sample

michaelosthege committed:
Merge branch 'sampling-docstrings' of https://github.com/michaelosthege/pymc3 into sampling-docstrings

michaelosthege commented:
@ColCarroll Thanks for your review!
I addressed your feedback and also added type hints and docstring to `_mp_sample`.
I also took out the creation of the unused `MultiTrace` already after checking that the `__init__` did not do something weird.

rpgoldman commented:
@michaelosthege  I botched some of these docstrings by replacing " : " in the parameters and return with ": " (i.e., dropping the space before the colon).
For some reason, when I run sphinx locally, it expects the incorrect spacing, instead of the one numpy dictates.  Sorry about that: mind fixing it throughout the file when you commit this?

michaelosthege committed:
remove spaces before colons in docstrings to make sphinx happy

rpgoldman commented:
@michaelosthege Thanks!
